### PR TITLE
[Trellis]Qwen3.5-35B; Weight Conversion and Raiden Weight Sync Integration

### PR DESCRIPTION
**Pull request by YixuanWang-99**
## Description

Integrates **Raiden weight synchronization** into `MaxTextTrainingEngine` for distributed RL post-training (Trellis / GRPO). Introduces trainer-side target-free weight conversion, MoE kernel layout alignment, hybrid KV cache mapping, and aggressive host memory reclamation during device-to-host staging.

**Companion Tunix PR**: [google/tunix#2083](https://github.com/google/tunix/pull/2083)

---

## Key Changes
- **Target-Free Weight Conversion (`weight_converter.py`, `convert_utils.py`)**: Synthesizes rollout target shapes and keys directly from trainer parameter state—unrolling scanned layers and prefusing MoE gate/up weights (`wi_0` + `wi_1` $\to$ `wi`) without requiring rollout engine `target_state`.
- **MoE 128-Lane Chunking & Padding (`convert_utils.py`, `moe_padding.py`)**: Implements 128-element lane interleaving (`TPU_V5P_SUBCORE_LANE_SIZE = 128`) for fused MoE weights to satisfy TPU GMM v2 kernel requirements, and dynamically pads intermediate dimensions.
- **Hybrid KV Cache Mapping (`hybrid_cache_utils.py`, `decoders.py`, `nnx_decoders.py`)**: Maps physical layer names to KV cache slot indices for hybrid architectures (e.g., Qwen 3.5 GDN recurrent + attention layers) and supports inhomogeneous layer unrolling.
- **Host Memory Reclamation & Pathways FFI (`maxtext_engine.py`)**: Adds `reclaim_host_memory()` (`gc.collect()` + `malloc_trim(0)`) and `release_host_arrays()` to eliminate host OOMs during D2H transfer; adds Pathways FFI support (`RAIDEN_USE_FFI`).
- **Drift Protection & Diagnostics**: Adds `cross_repo_drift_test.py` to guard against parameter/contract drift with Tunix, and replaces silent failure fallbacks with fail-fast exceptions.

---

## Testing
- **Unit Tests**: Added `cross_repo_drift_test.py`, `convert_utils_test.py`, `weight_converter_test.py`, `raiden_unscan_test.py`, `vllm_hybrid_cache_test.py`, and `prepare_weight_sync_test.py`.
- **E2E Verification**: Verified distributed GRPO training on Cloud TPU v5p (2×2×2 trainer, 2×2×1 rollout).  [logs](https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.cluster_name%3D%22bodaborg-v5p-nap%22%0Aresource.labels.pod_name%3D~%22yixuannw-rd-35b-roll.*%22%0Aresource.labels.container_name%3D%22main%22;cursorTimestamp=2026-09-08T19:37:58.048167180Z;customDuration=today?project=cloud-tpu-shared-capacity&e=13803378&mods=logs_tg_prod)

---

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/training_engine/maxtext_engine.py [hunk 1]
```diff
@@ -399,6 +416,29 @@ def __init__(
     self._metrics_recorder = metrics_module.MetricsRecorder()
     self._throttler = inflight_throttler.InflightThrottler(config=self._config)
     self._raiden_sync: Any = None
+    self._last_staged_step: Optional[int] = None
+    self._staged_metadata: Any = None
+    self._use_weight_converter = bool(
+        getattr(self._config, "use_weight_converter", False)
```
Comment (igorts-git) [inline src/maxtext/training_engine/maxtext_engine.py:422]: I am always very suspicious when AI Agents add "getattr" or "hasattr", it is usually an indications that they are referencing an old implementation or hallucinating APIs that don't exist. Our config either has a `use_weight_converter` flag or it does not. Using getattr is error-prone in case of misspelling or flag name change.
AI agents also tend to add those statements because they are lazy to update the configs in the unit tests.

File: src/maxtext/training_engine/maxtext_engine.py [hunk 2]
```diff
@@ -1482,33 +1553,55 @@ def prepare_weight_sync(
             "compatible tpu_raiden_jax wheel with FFI support is installed."
         )
 
+      raiden_ffi_env = os.environ.get("RAIDEN_USE_FFI")
```
Comment (igorts-git) [inline src/maxtext/training_engine/maxtext_engine.py:1556]: Why do we ever want to override this? IIUC, for any reasonable large model we **must** use FFI or we will OOM.

File: src/maxtext/training_engine/maxtext_engine.py [hunk 3]
```diff
@@ -1482,33 +1553,55 @@ def prepare_weight_sync(
...
+        raiden_ffi_env = os.environ.get("USE_RAIDEN_FFI", "false")
+      use_raiden_ffi = raiden_ffi_env.lower() in ("1", "true", "yes")
+      host_stage = is_pathways and not use_raiden_ffi
+
       if self._raiden_sync is None:
         self._raiden_sync = raiden_synchronizer.RaidenSynchronizer(
             job_name="trainer",
             worker_index=jax.process_index(),
             auto_h2d=False,
+            host_stage=host_stage,
```
Comment (igorts-git) [inline src/maxtext/training_engine/maxtext_engine.py:1567]: IIRC, "host_stage" was removed from the API in the last week. Do we have to add it back or is it a hallunication?

**Comment by YixuanWang-99** [comment]
Thanks everyone for the review feedback!

Per the feedback on PR size and scope, this PR is being split into a stacked chain of focused, reviewable PRs targeting `main`:

### Stacked PR Chain:
1. **#5166 (M7)**: Cross-repo drift guard & FFI resolution matrix tests (`tests/post_training/unit/cross_repo_drift_test.py`)
   - Test-only drift guard verifying cross-repo interface compatibility and FFI transport resolution across platforms.
2. **#5167 (M1)**: Inhomogeneous layer cycle support in `raiden_unscan` (`raiden_unscan.py` + tests)
   - Supports `cycle_interval` in `unscan_layers`. Fixes `slot is None` with a descriptive `ValueError` naming the key and cycle-slot prefix when `cycle_interval > 1`.
3. **#5168 (M2)**: MoE 128-lane layout and padding for TPU GMM kernels (`convert_utils.py`, `moe_padding.py`, `qwen35_moe.py`, `model_creation_utils.py`)
   - Implements 128-lane interleaving and padding matching TPU GMM layout. Removes dead `scan_fused_axis` parameter.
   - Pairs with Tunix #2150 (T5).
4. **#5169 (M3)**: Target-free streaming weight conversion for vLLM (`weight_converter.py`, `validate_converter.py` + tests)
   - Directly converts MaxText weights to target formats and streams per parameter to minimize host peak RSS.
5. **#5170 (M5)**: Shared expert gate `weight_dtype` fix (`qwen3.py`)
   - Explicitly passes `weight_dtype=cfg.weight_dtype` to `Qwen3NextSparseMoeBlock.shared_expert_gate`.
6. **#5171 (M4)**: TrainingEngine Raiden FFI integration (`maxtext_engine.py`, `types.py`, `prepare_weight_sync_test.py`)
   - **Addresses review comments**: Replaces `getattr` on config attributes with direct typed access (@igorts); defines `use_raiden_ffi: Optional[bool]` on config types.
   - Replaces ad-hoc env-var reading with Tunix's centralized `resolve_use_ffi` and `normalize_host_stage`.
   - Single synchronizer instance and rebind optimization.
   - Pairs with Tunix #2147 (T2a).
7. **#5172 (M6)**: Rollout path cleanup (`maxtext_vllm_rollout.py`, `train_rl.py`)
   - Net −354 lines: deletes legacy rollout duplication superseded by M3/M4.

### Notes on review comments & dropped duplicates:
- **Duplicate hybrid cache KV reordering**: Dropped in favor of #5139 (@sierraisland). Reverted `hybrid_cache_utils.py`, `adapter.py`, and decoders to decouple cleanly.
- `yixuann-debug-raiden` remains alive as the integration branch while landing the stacked PRs.
